### PR TITLE
First attempt at JOIN to Rel-Type

### PR DESCRIPTION
**Pull request by jexp**
* For foreign key column style (no JOIN table)

#18

Would love some conceptual feedback, see the TCK file for the current handling

The next one with the JOIN table would need to consider two connected JOINs not 100% sure how to get that other/next join from the current Table<?> or if we would have to do that a level up.

### CODE REVIEW

**Comment by lukaseder** [comment]
> Would love some conceptual feedback, see the TCK file for the current handling

Sure, I provided some review, but specific questions are good as well. I'm not sure yet where all your doubts lie, (not knowing too much about Cypher, and especially, common patterns and idioms when writing Cypher)

**Comment by jexp** [comment]
I'll add a few comments in the code.

File: src/main/java/org/neo4j/sql2cypher/Translator.java [hunk 1]
```diff
@@ -461,15 +448,55 @@ else if (c instanceof QOM.IsNotNull e) {
 		}
 	}
 
-	private Node node(Table<?> t) {
+	private PatternElement resolveTableOrJoin(Table<?> t) {
+		if (t instanceof QOM.Join<?> ta && ta.$on() instanceof QOM.Eq<?> eq) {
```
Comment (jexp) [inline src/main/java/org/neo4j/sql2cypher/Translator.java:452]: @lukaseder is this the best way to determine a simple equals join?

File: src/main/java/org/neo4j/sql2cypher/Translator.java [hunk 2]
```diff
@@ -461,15 +448,55 @@ else if (c instanceof QOM.IsNotNull e) {
...
+			var relType = eq.$arg1().getQualifiedName().last();
+			Table<?> t1 = ta.$table1();
+			Table<?> t2 = ta.$table2();
+			System.out.println("t1 = " + Arrays.toString(t1.getQualifiedName().getName()));
+			System.out.println("t2 = " + Arrays.toString(t2.getQualifiedName().getName()));
+			if (/*mightBeNodeLabel(t1.$name().first()) &&*/ resolveTableOrJoin(t1) instanceof Node from
```
Comment (jexp) [inline src/main/java/org/neo4j/sql2cypher/Translator.java:458]: Is there any way to structually determine join tables vs. entity tables?
And how do I access the actual table name, not the alias?

File: src/main/java/org/neo4j/sql2cypher/Translator.java [hunk 3]
```diff
@@ -461,15 +448,55 @@ else if (c instanceof QOM.IsNotNull e) {
...
+					/*&& mightBeNodeLabel(t2.$name().first())*/ && resolveTableOrJoin(t2) instanceof Node to
+					&& mightBeRelationshipType(relType) // upper-case-column-name
+			) {
+				// todo check against primary key
+				// var targetIsId = eq.$arg2().getQualifiedName().last().equals("id");
```
Comment (jexp) [inline src/main/java/org/neo4j/sql2cypher/Translator.java:463]: is there a better way of representing "id" of the target, not sure (also not sure how well supported something else would be in tools, like `t.ROWID` or such)

File: src/main/java/org/neo4j/sql2cypher/Translator.java [hunk 4]
```diff
@@ -461,15 +448,55 @@ else if (c instanceof QOM.IsNotNull e) {
...
+				return from.relationshipTo(to, relType);
+			}
+			else {
+				throw unsupported(ta);
+			}
+		}
+
```
Comment (jexp) [inline src/main/java/org/neo4j/sql2cypher/Translator.java:470]: Case number 2)

How would I structurally (via instanceof) determine here the `t1 JOIN join_table as jt ON (t1.id = jt.t1_id) JOIN t2 ON (jt.t2_id = t2.id)` equivalent ?

**Comment by jexp** [comment]
@lukaseder unrelated FYI: not sure if the build failure trying to compile jOOQ, was only a temporary glitch in the snapshot?
https://github.com/neo4j-contrib/sql2cypher/actions/runs/3991841532/jobs/6847082074

**Comment by lukaseder** [comment]
> @lukaseder unrelated FYI: not sure if the build failure trying to compile jOOQ, was only a temporary glitch in the snapshot?
> https://github.com/neo4j-contrib/sql2cypher/actions/runs/3991841532/jobs/6847082074

https://github.com/neo4j-contrib/sql2cypher/issues/27

**Comment by jexp** [comment]
@lukaseder can you have a look at my code-inline questions?
I'd be happy to merge this as a WIP state, but want to make sure that I'm not doing stupid things.

**Comment by jexp** [comment]
@lukaseder I meant on these:
Bildschirmfoto 2023-01-25 um 18 34 10
Bildschirmfoto 2023-01-25 um 18 34 03

**Comment by jexp** [comment]
@lukaseder GitHub is odd at that, it only sends them immediately if you use "Add single comment" and otherwise it adds/collects them to your "Review" and only publishes them when you finish your "Review" in the top-right box.

**Comment by jexp** [comment]
Thanks for the answers, I'll go through them.

**Comment by lukaseder** [comment]
Ugh, I didn't realise I had started a formal "review", sorry

**Comment by jexp** [comment]
It auto starts a review if you don’t use the single comment button.

File: src/main/java/org/neo4j/sql2cypher/Translator.java (hunk 1 above)
Comment (jexp) [reply src/main/java/org/neo4j/sql2cypher/Translator.java:558]: yeah, perhaps we can add some failing tests, then for option 1, we can move the predicates to the WHERE clause.
NOT DISTINCT FROM - need to figure out how to express that in cypher.

Multi-key pair equalities should work the same as single ones (just map to multiple rels), would be probably good to filter for equalities, process them all and add the remaining predicates to the WHERE.

In terms of rare syntax we can just fail with "not-implemented" for the time being.

File: src/main/java/org/neo4j/sql2cypher/Translator.java (hunk 2 above)
Comment (jexp) [reply src/main/java/org/neo4j/sql2cypher/Translator.java]: join table in my case is what represents the relationship, i.e. the many-to-many table with 2 foreign keys and a bunch of other columns (or no other columns) in many cases these don't have a primary key but are identified by the two foreign keys.

Entity tables have mostly a primary key, some data columns and any number of FK's 0..n

File: src/main/java/org/neo4j/sql2cypher/Translator.java [hunk 5]
```diff
@@ -461,15 +448,55 @@ else if (c instanceof QOM.IsNotNull e) {
...
+	private static boolean mightBeRelationshipType(String relType) {
+		System.out.println("relType = " + relType);
+		return relType.matches("^[A-Z]+(_[A-Z]+)*$");
+	}
+
+	// via naming convention or via mapping configuration
+	private static boolean mightBeNodeLabel(String label) {
+		System.out.println("label = " + label);
+		return label.matches("^[A-Z][a-z]+([A-Z][a-z]+)*$");
+	}
```
[the remark this answers is not in the document]
Comment (jexp) [reply src/main/java/org/neo4j/sql2cypher/Translator.java:602]: Yes exactly, but right now we haven't wired up a proper Meta resolver. And I don't know what that would look like.
So perhaps a name based resolver is good enough for now.

**Comment by jexp** [comment]
Merging this now to make progress, next up is a multi-way join.

**Pull request merged**